packages/protocol/contracts/dapis/RrpDapiServer.sol

```diff
@@ -0,0 +1,178 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.4;
+
+import "../AirnodeRrp.sol";
+import "./reducers/MeanMedianHybrid.sol";
+
+contract RrpDapiServer is MeanMedianHybrid {
+    struct Dapi {
+        uint256 noResponsesToReduce;
+        int256 toleranceInPercentages;
+        uint256 requesterIndex;
+        bytes32[] templateIds;
+        address[] designatedWallets;
+        address reduceAddress;
+        bytes4 reduceFunctionId;
+        mapping(uint256 => int256[]) requestIndexToResponses;
+        mapping(uint256 => uint256) requestIndexToResponsesLength;
+        uint64 nextRequestIndex;
+        uint64 requestIndexResetCount;
+        address requestIndexResetter;
+        }
+    
+    struct DapiRequestIdentifiers {
+        bytes16 dapiId;
+        uint64 dapiRequestIndex;
+        uint64 dapiRequestIndexResetCountAtRequestTime;
+        }
+    
+    AirnodeRrp public airnodeRrp;
+    mapping(bytes16 => Dapi) private dapis;
+    uint256 public nextDapiIndex = 1;
+    mapping(bytes32 => DapiRequestIdentifiers) private airnodeRequestIdToDapiRequestIdentifiers;
+
+    constructor(address _airnodeRrp) {
+        airnodeRrp = AirnodeRrp(_airnodeRrp);
+    }
+
+    function registerDapi(
+        uint256 noResponsesToReduce,
+        int256 toleranceInPercentages,
+        uint256 requesterIndex,
+        bytes32[] calldata templateIds,
+        address[] calldata designatedWallets,
+        address reduceAddress,
+        bytes4 reduceFunctionId,
+        address requestIndexResetter
+        )
+        external
+        returns (bytes16 dapiId)
+    {
+      require(
+          templateIds.length == designatedWallets.length,
+          "Parameter lengths do not match"
+          );
+      require(
+          noResponsesToReduce <= templateIds.length && noResponsesToReduce != 0,
+          "Invalid number of responses to reduce"
+          );
+      dapiId = bytes16(keccak256(abi.encodePacked(
+          noResponsesToReduce,
+          toleranceInPercentages,
+          requesterIndex,
+          templateIds,
+          designatedWallets,
+          reduceAddress,
+          reduceFunctionId,
+          requestIndexResetter
+          )));
```

Comment:
I'm pretty sure this is very cheap. Note that I'm using half of the hash for identification because I want to pack into a 32 byte struct.

Reply:
What do you mean by "I'm using half of the hash for identification"?

Reply:
`keccak256()` returns a `bytes32`. I typecast it as `bytes16`, essentially throwing half of it away. Now it is a lot more likely for dAPI IDs to collide, but still unlikely enough.